### PR TITLE
Toggle switch control (ucr)

### PR DESCRIPTION
**Pull request by SusanRatiLane**
This is a toggle switch component. Functionally, it behaves the same as a checkbox: the ON state is equivalent to checked.

Since iOS does not have a checkbox and only offers toggle switches, this component paves the way for an on/off component that is shared by both Android and iOS.

The toggle switch is a 2-color control, which means that there are 4 configurable colors, 2 for on and 2 for off.

### CODE REVIEW

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+  /**
+   * Default LostFocus event handler.
+   */
+  @SimpleEvent
+  public void LostFocus() {
+    EventDispatcher.dispatchEvent(this, "LostFocus");
+  }
+
+  /**
+   * Returns the checkbox's background color as an alpha-red-green-blue
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: The description might be too detailed by saying how the color is encoded.

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+   * @return  background RGB color with alpha
+   */
+  @SimpleProperty(
+      category = PropertyCategory.APPEARANCE)
+  public int BackgroundColor() {
+    return backgroundColor;
+  }
+
+  /**
+   * Specifies the checkbox's background color as an alpha-red-green-blue
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: Ditto color concern above.

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+    if (argb != Component.COLOR_DEFAULT) {
+      TextViewUtil.setBackgroundColor(view, argb);
+    } else {
+      TextViewUtil.setBackgroundColor(view, Component.COLOR_NONE);
+    }
+  }
+
+
+  /**
+   * Returns the checkbox's thumb color as an alpha-red-green-blue
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: Ditto color concern above.

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 4]
```diff
@@ -0,0 +1,473 @@
...
+   *
+   * @return  thumb RGB color with alpha
+   */
+  @SimpleProperty(category = PropertyCategory.APPEARANCE, userVisible = false)
+  public int ThumbColorActive() {
+    return thumbColorActive;
+  }
+
+  /**
+   * Specifies the checkbox's thumb color as an alpha-red-green-blue
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: Ditto color concern above.

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_COLOR,
+          defaultValue = Component.DEFAULT_VALUE_COLOR_WHITE)
+  @SimpleProperty
+  public void ThumbColorActive(int argb) {
+    thumbColorActive = argb;
+    DrawableCompat.setTintList(view.getThumbDrawable(), createSwitchColors(argb, thumbColorInactive));
+    view.invalidate();
+  }
+
+  @SimpleProperty(category = PropertyCategory.APPEARANCE, userVisible = true)
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: ThumbColorInactive needs a description.

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+   * @return  {@code true} indicates enabled, {@code false} disabled
+   */
+  @SimpleProperty(
+      category = PropertyCategory.BEHAVIOR)
+  public boolean Enabled() {
+    return view.isEnabled();
+  }
+
+  /**
+   * Specifies whether the checkbox should be active and clickable.
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 7]
```diff
@@ -0,0 +1,473 @@
...
+   */
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_BOOLEAN,
+      defaultValue = "True")
+  @SimpleProperty
+  public void Enabled(boolean enabled) {
+    TextViewUtil.setEnabled(view, enabled);
+  }
+
+  /**
+   * Returns true if the checkbox's text should be bold.
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 8]
```diff
@@ -0,0 +1,473 @@
...
+   */
+  @SimpleProperty(
+      category = PropertyCategory.APPEARANCE,
+      userVisible = false)
+  public boolean FontBold() {
+    return bold;
+  }
+
+  /**
+   * Specifies whether the checkbox's text should be bold.
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 9]
```diff
@@ -0,0 +1,473 @@
...
+      defaultValue = "False")
+  @SimpleProperty(
+      userVisible = false)
+  public void FontBold(boolean bold) {
+    this.bold = bold;
+    TextViewUtil.setFontTypeface(view, fontTypeface, bold, italic);
+  }
+
+  /**
+   * Returns true if the checkbox's text should be italic.
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 10]
```diff
@@ -0,0 +1,473 @@
...
+   */
+  @SimpleProperty(
+      category = PropertyCategory.APPEARANCE,
+      userVisible = false)
+  public boolean FontItalic() {
+    return italic;
+  }
+
+  /**
+   * Specifies whether the checkbox's text should be italic.
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 11]
```diff
@@ -0,0 +1,473 @@
...
+      defaultValue = "False")
+  @SimpleProperty(
+      userVisible = false)
+  public void FontItalic(boolean italic) {
+    this.italic = italic;
+    TextViewUtil.setFontTypeface(view, fontTypeface, bold, italic);
+  }
+
+  /**
+   * Returns the checkbox's text's font size, measured in sp(scale-independent pixels).
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+   * @return  font size in sp (scale-independent pixels)
+   */
+  @SimpleProperty(
+      category = PropertyCategory.APPEARANCE)
+  public float FontSize() {
+    return TextViewUtil.getFontSize(view, container.$context());
+  }
+
+  /**
+   * Specifies the checkbox's text's font size, measured in sp(scale-independent pixels).
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+   */
+  @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_NON_NEGATIVE_FLOAT,
+      defaultValue = Component.FONT_DEFAULT_SIZE + "")
+  @SimpleProperty
+  public void FontSize(float size) {
+    TextViewUtil.setFontSize(view, size);
+  }
+
+  /**
+   * Returns the checkbox's text's font face as default, serif, sans
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: checkbox => switch

**Comment by ewpatton** [comment]
Two more thoughts:

1. This should have some unit tests. You can take a look at how I implemented the tests for Maps.
2. For the Switch, I think that an On/Off paradigm is better than Checked as a property. We might have the property On instead of Checked, and optionally an Off which is just syntactic sugar for not On

**Comment by SusanRatiLane** [comment]
I believe all feedback is addressed. This is ready for review.

A good look at unit tests would be appreciated.

I have not added an Off property. I would normally use isOn, but that doesn't seem to be the naming convention here.

**Comment by SusanRatiLane** [comment]
Whoops. Forgot to mention that the On/Off property is now in the PR.

File: appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockSwitch.java [hunk 14]
```diff
@@ -0,0 +1,213 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2011-2018 MIT, All rights reserved
```
Comment (ewpatton) [inline appinventor/appengine/src/com/google/appinventor/client/editor/simple/components/MockSwitch.java]: Just need 2018 for the Copyright year here.

File: appinventor/components/src/com/google/appinventor/components/runtime/CheckBox.java [hunk 15]
```diff
@@ -334,7 +47,7 @@ public void TextColor(int argb) {
    * @return  {@code true} indicates checked, {@code false} unchecked
    */
   @SimpleProperty(
-      category = PropertyCategory.BEHAVIOR)
+          category = PropertyCategory.BEHAVIOR)
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/CheckBox.java]: Undo the whitespace change here.

File: appinventor/components/src/com/google/appinventor/components/runtime/CheckBox.java [hunk 16]
```diff
@@ -345,26 +58,11 @@ public boolean Checked() {
    * @param value  {@code true} indicates checked, {@code false} unchecked
    */
   @DesignerProperty(editorType = PropertyTypeConstants.PROPERTY_TYPE_BOOLEAN,
-      defaultValue = "False")
+          defaultValue = "False")
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/CheckBox.java]: Ditto whitespace.

File: appinventor/components/src/com/google/appinventor/components/runtime/Switch.java [hunk 17]
```diff
@@ -0,0 +1,180 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2011-2018 MIT, All rights reserved
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/Switch.java]: Just need 2018 here.

File: appinventor/components/tests/com/google/appinventor/components/runtime/CheckboxTest.java [hunk 18]
```diff
@@ -0,0 +1,61 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test class for Switch component.
+ *
+ * @author srlane@mit.edu (Susan Rati Lane)
+ */
+@Config(shadows = {ShadowMapView.class, ShadowMapTileModuleProviderBase.class})
```
Comment (ewpatton) [inline appinventor/components/tests/com/google/appinventor/components/runtime/CheckboxTest.java]: You shouldn't need these shadows here as they are specific to testing Map-related components.

File: appinventor/components/tests/com/google/appinventor/components/runtime/SwitchTest.java [hunk 19]
```diff
@@ -0,0 +1,83 @@
...
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * Test class for Switch component.
+ *
+ * @author srlane@mit.edu (Susan Rati Lane)
+ */
+@Config(shadows = {ShadowMapView.class, ShadowMapTileModuleProviderBase.class})
```
Comment (ewpatton) [inline appinventor/components/tests/com/google/appinventor/components/runtime/SwitchTest.java]: Ditto previous comment about removing unnecessary shadows.

File: appinventor/components/tests/com/google/appinventor/components/runtime/ToggleTestBase.java [hunk 20]
```diff
@@ -0,0 +1,54 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright © 2017-2018 Massachusetts Institute of Technology, All rights reserved.
```
Comment (ewpatton) [inline appinventor/components/tests/com/google/appinventor/components/runtime/ToggleTestBase.java]: Only 2018

File: appinventor/components/tests/com/google/appinventor/components/runtime/SwitchTest.java [hunk 21]
```diff
@@ -0,0 +1,83 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright © 2017-2018 Massachusetts Institute of Technology, All rights reserved.
```
Comment (ewpatton) [inline appinventor/components/tests/com/google/appinventor/components/runtime/SwitchTest.java]: Only 2018

File: appinventor/components/tests/com/google/appinventor/components/runtime/CheckboxTest.java [hunk 22]
```diff
@@ -0,0 +1,61 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright © 2017-2018 Massachusetts Institute of Technology, All rights reserved.
```
Comment (ewpatton) [inline appinventor/components/tests/com/google/appinventor/components/runtime/CheckboxTest.java]: Only 2018

File: appinventor/components/src/com/google/appinventor/components/runtime/ToggleBase.java [hunk 23]
```diff
@@ -0,0 +1,346 @@
+// -*- mode: java; c-basic-offset: 2; -*-
+// Copyright 2011-2018 MIT, All rights reserved
```
Comment (ewpatton) [inline appinventor/components/src/com/google/appinventor/components/runtime/ToggleBase.java]: Given that most of the code here came from Google (via Checkbox.java), I think we should retain the Google copyright here.

File: appinventor/components/tests/com/google/appinventor/components/runtime/ToggleTestBase.java [hunk 24]
```diff
@@ -0,0 +1,54 @@
...
+  ToggleBase aToggle;
+
+  @Test
+  public void testToggleDefaults() {
+    assertEquals("Expected default background color NONE",Component.COLOR_NONE, aToggle.BackgroundColor());
+    assertEquals("Expected default Typeface " + Component.TYPEFACE_DEFAULT,Component.TYPEFACE_DEFAULT, aToggle.FontTypeface());
+    assertTrue(aToggle.Enabled());
+    assertEquals(Component.FONT_DEFAULT_SIZE, aToggle.FontSize(), 0.0);
+    assertEquals("Expected default Text Color " + Component.COLOR_DEFAULT, Component.COLOR_DEFAULT, aToggle.TextColor());
+    assertFalse(aToggle.Checked());
```
Comment (ewpatton) [inline appinventor/components/tests/com/google/appinventor/components/runtime/ToggleTestBase.java]: Test fails to compile here.

Comment (SusanRatiLane) [reply appinventor/components/tests/com/google/appinventor/components/runtime/ToggleTestBase.java]: Compile issue should be fixed.

**Comment by SusanRatiLane** [comment]
Issues should be addressed. This is ready for review.

**Comment by SusanRatiLane** [comment]
The app now pops up a warning and closes if the API is less than 14 (Ice Cream Sandwich). This is ready for review.

**Comment by SusanRatiLane** [comment]
Whoops. Docco should now be there.

**Pull request merged**